File: app/src/main/java/com/routesme/taxi_screen/java/View/Login/LoginScreen.java
package com.routesme.taxi_screen.java.View.Login;

import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.FragmentActivity;
import androidx.lifecycle.LifecycleOwner;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProviders;
import android.app.Dialog;
import android.app.ProgressDialog;
import android.content.Intent;
import android.graphics.Color;
import android.graphics.drawable.ColorDrawable;
import android.os.Bundle;
import android.provider.Settings;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.RelativeLayout;
import android.widget.TextView;
import com.andrognito.patternlockview.PatternLockView;
import com.andrognito.patternlockview.listener.PatternLockViewListener;
import com.andrognito.patternlockview.utils.PatternLockUtils;
import com.routesme.taxi_screen.java.Class.App;
import com.routesme.taxi_screen.java.Class.Operations;
import com.routesme.taxi_screen.java.Model.AuthCredentialsViewModel;
import com.routesme.taxi_screen.kotlin.Model.AuthCredentials;
import com.routesme.taxi_screen.kotlin.Model.AuthCredentialsError;
import com.routesme.taxiscreen.R;
import com.scottyab.showhidepasswordedittext.ShowHidePasswordEditText;
import java.util.List;

public class LoginScreen extends AppCompatActivity implements View.OnClickListener {

    private Operations operations;
    private App app;
    private ImageView btnOpenLoginScreen;
    private RelativeLayout loginLayout;
    private long PressedTime;
    private int clickTimes = 0;
    //For Login Layout...
    private String userName = null, password = null;
    View technical_login_screen ;
    private AuthCredentialsViewModel authCredentialsViewModel;
    private Button btn_next;
    private LinearLayout btn_learnMore;
    private EditText userName_et;
    private ShowHidePasswordEditText password_et;
    private TextView userName_error_tv, password_error_tv;
    private ProgressDialog dialog;
    private String final_pattern = "";
    private Dialog exitPatternDialog;
    private PatternLockView pattern_exitApp;
    private ImageView openPattern;
    private String patternPassword;




    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.login_screen);
        overridePendingTransition(android.R.anim.fade_in, android.R.anim.fade_out);

        initialize();
    }


    private void initialize() {
        operations = new Operations(this);
        app = (App) getApplicationContext();
        btnOpenLoginScreen = findViewById(R.id.btnOpenLoginScreen);
        btnOpenLoginScreen.setOnClickListener(this);
        loginLayout = findViewById(R.id.loginLayout);
        openPattern = findViewById(R.id.openPattern);
        openPattern.setOnClickListener(this);
        initializeLoginLayout();
        openLoginLayout(app.isNewLogin());
    }

    private void initializeLoginLayout() {
        technical_login_screen = findViewById(R.id.technical_login_screen);
        dialog = new ProgressDialog(this);
        dialog.setMessage("Please Wait...");
        dialog.setCancelable(false);

        btn_next = technical_login_screen.findViewById(R.id.btn_next);
        btn_next.setOnClickListener(this);
        operations.enableNextButton(btn_next,true);
        btn_learnMore = technical_login_screen.findViewById(R.id.btn_learnMore);
        btn_learnMore.setOnClickListener(this);

        userName_et = technical_login_screen.findViewById(R.id.userName_et);
        userName_error_tv = technical_login_screen.findViewById(R.id.userName_error_tv);
        password_et = technical_login_screen.findViewById(R.id.password_et);
        password_error_tv = technical_login_screen.findViewById(R.id.password_error_tv);
        editTextListener();
    }


    @Override
    public void onClick(View v) {
        switch (v.getId()){
            case R.id.btnOpenLoginScreen:
                showLoginView();
                break;
            case R.id.btn_next:
                operations.enableNextButton(btn_next,false);
                saveAuthCredentials();
                openTaxiInformationScreen();
                break;
            case R.id.btn_learnMore:
                saveAuthCredentials();
                openLearnMoreScreen();
                break;

            case R.id.openPattern:
                openPatternDialog();
                break;
        }
    }

    private void openPatternDialog() {
            clickTimes++;
            if (PressedTime + 1000 > System.currentTimeMillis() && clickTimes >= 10) {
                readPatternFromTechnicalSupport();
                clickTimes = 0;
            }
            PressedTime = System.currentTimeMillis();
    }

    public void readPatternFromTechnicalSupport() {
        patternPassword = "2103678";
        showExitPatternDialog();

        pattern_exitApp.addPatternLockListener(new PatternLockViewListener() {
            @Override
            public void onStarted() {
            }
            @Override
            public void onProgress(List<PatternLockView.Dot> progressPattern) {
            }
            @Override
            public void onComplete(List<PatternLockView.Dot> pattern) {
                final_pattern = PatternLockUtils.patternToString(pattern_exitApp, pattern);
                if (final_pattern.equals(patternPassword)) {
                    openSettings();
                } else {
                    pattern_exitApp.clearPattern();
                    exitPatternDialog.dismiss();
                }
            }
            @Override
            public void onCleared() {
            }
        });
    }

    private void showExitPatternDialog() {
            exitPatternDialog = new Dialog(this);
            exitPatternDialog.setContentView(R.layout.exit_pattern_dialog);
            pattern_exitApp = exitPatternDialog.findViewById(R.id.admin_verification_pattern);
            exitPatternDialog.getWindow().setBackgroundDrawable(new ColorDrawable(Color.TRANSPARENT));
            exitPatternDialog.show();
            exitPatternDialog.setCancelable(false);
    }

    private void openSettings() {
            if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.LOLLIPOP) {
                final Intent intent = new Intent(Settings.ACTION_HOME_SETTINGS);
                startActivity(intent);
            } else {
                final Intent intent = new Intent(Settings.ACTION_SETTINGS);
                startActivity(intent);
            }
            finish();
            System.exit(0);
    }

    private void openTaxiInformationScreen() {
        String userName = userName_et.getText().toString().trim();
        String password = password_et.getText().toString().trim();
        AuthCredentials authCredentials = new AuthCredentials(userName, password);
        authCredentialsViewModel = ViewModelProviders.of((FragmentActivity) this).get(AuthCredentialsViewModel.class);
        authCredentialsViewModel.getToken(authCredentials,this, dialog).observe((LifecycleOwner) this, new Observer<List<AuthCredentialsError>>() {
            @Override
            public void onChanged(List<AuthCredentialsError> authCredentialsErrors) {
                for (int e = 0 ; e < authCredentialsErrors.size() ; e++ ){
                    if (authCredentialsErrors.get(e).getErrorNumber() == 1 || authCredentialsErrors.get(e).getErrorNumber() == 2){
                        showErrorMessage(authCredentialsErrors.get(e).getErrorNumber(),authCredentialsErrors.get(e).getErrorMessage(),true);
                    }
                }
            }
        });
    }

    private void showLoginView() {
            clickTimes++;
            if (PressedTime + 1000 > System.currentTimeMillis() && clickTimes >= 10) {
                openLoginLayout(true);
            }
            PressedTime = System.currentTimeMillis();
    }

    private void editTextListener() {
        userName_et.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) { }
            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                showErrorMessage(1,"",false);
                operations.enableNextButton(btn_next,true);
            }
            @Override
            public void afterTextChanged(Editable s) { }
        });
        password_et.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) { }
            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                showErrorMessage(2,"",false);
                operations.enableNextButton(btn_next,true);
            }
            @Override
            public void afterTextChanged(Editable s) { }
        });
    }

    private void showErrorMessage(int errorId , String errorStr, boolean show){
        EditText editText = null;
        TextView textView = null;
        switch (errorId){
            case 1 :
                editText = userName_et;
                textView = userName_error_tv;
                break;
            case 2 :
                editText = password_et;
                textView = password_error_tv;
                break;
        }
        if (show){
            editText.setBackgroundResource(R.drawable.red_border);
            textView.setText("* " + errorStr);
            textView.setVisibility(View.VISIBLE);
            return;
        }else {
            editText.setBackgroundResource(R.drawable.grey_border_edit_text);
            textView.setVisibility(View.INVISIBLE);
        }
    }
    private void openLearnMoreScreen() {
        startActivity(new Intent(this, LearnMoreScreen.class));
    }

    private void openLoginLayout(boolean show){
        if (show) {
            btnOpenLoginScreen.setVisibility(View.GONE);
            loginLayout.setVisibility(View.VISIBLE);
            userName = app.getTechnicalSupportUserName();
            password = app.getTechnicalSupportPassword();
            if (userName != null && !userName.isEmpty()){
                userName_et.setText(userName);
            }
            if (password != null && !password.isEmpty()){
                password_et.setText(password);
            }
        }else{
            btnOpenLoginScreen.setVisibility(View.VISIBLE);
            loginLayout.setVisibility(View.GONE);
        }
    }

    private void saveAuthCredentials(){
            String userName = userName_et.getText().toString().trim();
            String password = password_et.getText().toString().trim();
            app.setTechnicalSupportUserName(userName);
            app.setTechnicalSupportPassword(password);
            app.setNewLogin(true);
    }

}

File: app/src/main/java/com/routesme/taxi_screen/kotlin/View/LoginScreens/LoginScreen.kt
package com.routesme.taxi_screen.kotlin.View.LoginScreens

import android.annotation.SuppressLint
import android.app.Activity
import android.app.AlertDialog
import android.app.Dialog
import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.provider.Settings
import android.text.Editable
import android.text.TextWatcher
import android.view.View
import android.widget.EditText
import android.widget.TextView
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.Observer
import com.andrognito.patternlockview.PatternLockView
import com.andrognito.patternlockview.listener.PatternLockViewListener
import com.andrognito.patternlockview.utils.PatternLockUtils
import com.google.gson.Gson
import com.google.gson.JsonArray
import com.google.gson.JsonElement
import com.google.gson.reflect.TypeToken
import com.routesme.taxi_screen.java.View.Login.TaxiInformationScreen
import com.routesme.taxi_screen.kotlin.Class.AesBase64Wrapper
import com.routesme.taxi_screen.kotlin.Class.App
import com.routesme.taxi_screen.kotlin.Class.Operations
import com.routesme.taxi_screen.kotlin.Model.AuthCredentials
import com.routesme.taxi_screen.kotlin.Model.AuthCredentialsError
import com.routesme.taxi_screen.kotlin.Model.Token
import com.routesme.taxi_screen.kotlin.ViewModel.RoutesViewModel
import com.routesme.taxiscreen.R
import dmax.dialog.SpotsDialog
import kotlinx.android.synthetic.main.exit_pattern_dialog.*
import kotlinx.android.synthetic.main.login_screen.*
import kotlinx.android.synthetic.main.technical_login_layout.*
import kotlinx.android.synthetic.main.technical_login_layout.view.*
import kotlin.system.exitProcess

class LoginScreen : AppCompatActivity() {

    private var pressedTime: Long = 0
    private var clickTimes = 0
    private val app = App.instance
    private lateinit var authCredentials: AuthCredentials
    private lateinit var userName: String
    private lateinit var password: String
    private lateinit var dialog: AlertDialog

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.login_screen)
        overridePendingTransition(android.R.anim.fade_in, android.R.anim.fade_out)

        initialize()
    }

    private fun initialize() {
        dialogSetUp()
        openLoginLayout(app!!.isNewLogin)
        btnOpenLoginScreen.setOnClickListener { showLoginView() }
        openPattern.setOnClickListener { openPatternDialog() }
        technical_login_screen.btn_learnMore.setOnClickListener { openLearnMoreScreen() }
        technical_login_screen.btn_next.setOnClickListener { buttonNextClick() }
        editTextListener()
    }

    private fun editTextListener() {
        userName_et.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                showErrorMessage(1, "", false)
                Operations().enableNextButton(btn_next, true)
            }

            override fun afterTextChanged(s: Editable) {}
        })
        password_et.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                showErrorMessage(2, "", false)
                Operations().enableNextButton(btn_next, true)
            }

            override fun afterTextChanged(s: Editable) {}
        })
    }

    private fun dialogSetUp() {
        dialog = SpotsDialog.Builder()
                .setContext(this)
                .setTheme(R.style.SpotsDialogStyle)
                .setCancelable(false).build()


    }

    private fun buttonNextClick() {
        Operations().enableNextButton(btn_next, false)
        saveAuthCredentials()
        login()
    }

    private fun login() {
        if (userNameValid() && passwordValid()) {
            dialog.show()
            val aesBase64Wrapper = AesBase64Wrapper(this)
            val authCredentials = AuthCredentials(aesBase64Wrapper.encryptAndEncode(userName), aesBase64Wrapper.encryptAndEncode(password))
            val model: RoutesViewModel by viewModels()
            model.getToken(authCredentials, dialog, this).observe(this, Observer<JsonElement> {
                if (it.isJsonObject) {
                    val accessToken = Gson().fromJson<Token>(it, Token::class.java).access_token
                    if (!accessToken.isNullOrEmpty()) {
                        saveDataIntoSharedPreference(accessToken)
                        openTaxiInformationScreen()
                    }
                } else if (it.isJsonArray) {
                    val authErrors = Gson().fromJson<List<AuthCredentialsError>>(it as JsonArray?, object : TypeToken<List<AuthCredentialsError?>?>() {}.type)
                    if (authErrors != null) {
                        for (e in authErrors.indices) {
                            if (authErrors[e].ErrorNumber == 1 || authErrors[e].ErrorNumber == 2) {
                                showErrorMessage(authErrors[e].ErrorNumber, authErrors[e].ErrorMessage.toString(), true)
                            }
                        }
                    }
                }
            })
        }
    }

    private fun saveDataIntoSharedPreference(access_token: String) {
        val editor = getSharedPreferences("userData", Activity.MODE_PRIVATE).edit()
        editor.putString("tabToken", access_token).apply()
    }

    private fun openTaxiInformationScreen() {
        startActivity(Intent(this, TaxiInformationScreen::class.java))
        finish()
    }

    private fun userNameValid(): Boolean {
        return if (userName.isNotEmpty()) true
        else {
            showErrorMessage(1, "User Name Required", true)
            false
        }
    }

    private fun passwordValid(): Boolean {
        return when {
            password.isEmpty() -> {
                showErrorMessage(2, "Password Required", true); false
            }
            password.length < 6 -> {
                showErrorMessage(2, "Minimum Password is 6 digit", true); false
            }
            else -> true
        }
    }

    @SuppressLint("SetTextI18n")
    private fun showErrorMessage(errorId: Int, errorStr: String, show: Boolean) {
        var editText: EditText? = null
        var textView: TextView? = null
        when (errorId) {
            1 -> {
                editText = userName_et
                textView = userName_error_tv
            }
            2 -> {
                editText = password_et
                textView = password_error_tv
            }
        }
        if (show) {
            editText!!.setBackgroundResource(R.drawable.red_border)
            textView!!.text = "* $errorStr"
            textView.visibility = View.VISIBLE
            return
        } else {
            editText!!.setBackgroundResource(R.drawable.grey_border_edit_text)
            textView!!.visibility = View.INVISIBLE
        }
    }

    private fun openPatternDialog() {
        clickTimes++
        if (pressedTime + 1000 > System.currentTimeMillis() && clickTimes >= 10) {
            readPatternFromTechnicalSupport()
            clickTimes = 0
        }
        pressedTime = System.currentTimeMillis()
    }

    private fun readPatternFromTechnicalSupport() {
        val patternPassword = "2103678"
        val exitPatternDialog = Dialog(this)
        exitPatternDialog.setContentView(R.layout.exit_pattern_dialog)
        exitPatternDialog.show()
        exitPatternDialog.setCancelable(false)
        val patternExitApp = exitPatternDialog.admin_verification_pattern
        patternExitApp.addPatternLockListener(object : PatternLockViewListener {
            override fun onStarted() {}
            override fun onProgress(progressPattern: List<PatternLockView.Dot>) {}
            override fun onComplete(pattern: List<PatternLockView.Dot>) {
                val finalPattern = PatternLockUtils.patternToString(patternExitApp, pattern)
                if (finalPattern == patternPassword) {
                    openSettings()
                } else {
                    patternExitApp.clearPattern()
                    exitPatternDialog.dismiss()
                }
            }

            override fun onCleared() {}
        })
    }

    private fun openSettings() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            startActivity(Intent(Settings.ACTION_HOME_SETTINGS))
        } else {
            startActivity(Intent(Settings.ACTION_SETTINGS))
        }
        finish()
        exitProcess(0)
    }

    private fun showLoginView() {
        clickTimes++
        if (pressedTime + 1000 > System.currentTimeMillis() && clickTimes >= 10) {
            openLoginLayout(true)
            clickTimes = 0
        }
        pressedTime = System.currentTimeMillis()
    }

    private fun openLoginLayout(show: Boolean) {
        if (show) {
            btnOpenLoginScreen.visibility = View.GONE
            loginLayout.visibility = View.VISIBLE
            showSavedCredentials()
        } else {
            btnOpenLoginScreen.visibility = View.VISIBLE
            loginLayout.visibility = View.GONE
        }
    }

    private fun showSavedCredentials() {
        if (app?.authCredentials != null) {
            authCredentials = app.authCredentials!!
            userName = authCredentials.Username
            password = authCredentials.Password
            if (userName.isNotEmpty()) {
                userName_et.setText(userName)
            }
            if (password.isNotEmpty()) {
                password_et.setText(password)
            }
        }
    }

    private fun openLearnMoreScreen() {
        saveAuthCredentials()
        startActivity(Intent(this, LearnMoreScreen::class.java))
    }

    private fun saveAuthCredentials() {
        userName = userName_et.text.toString().trim()
        password = password_et.text.toString().trim()
        authCredentials = AuthCredentials(userName, password)
        app!!.authCredentials = authCredentials
        app.isNewLogin = true
    }
}